barman/compression.py

```diff
@@ -447,53 +597,200 @@ def with_suffix(self, basename):
+            if exclude is not None and member.path in exclude:
+                continue
+            if include is not None and member.path in include:
+                members_subset.append(member)
+            if include is None:
+                members_subset.append(member)
+        tar.extractall(path=destination, members=members_subset)
+
+
+class Compression(with_metaclass(ABCMeta, object)):
+    """
+    Abstract class meant to represent compression interface
+    """
+
+    @abstractproperty
+    def name(self):
+        """
+
+        :return:
+        """
+
+    @abstractproperty
+    def file_extension(self):
+        """
+
+        :return:
+        """
+
+    @abstractmethod
+    def uncompress(self, src, dst, exclude=None, include_args=None):
+        """
+
+        :param src: source file path without compression extension
+        :param dst: destination path
+        :param exclude: list of filepath in the archive to exclude from the extraction
+        :param include_args: list of filepath in the archive to extract.
+        :return:
+        """
+
+    @abstractmethod
+    def get_file_content(self, filename, archive):
+        """
+
+        :param filename: str file to search for in the archive (requires its full path within the archive)
+        :param archive: str archive path/name without extension
+        :return: string content
+        """
+
+
+class GZipCompression(Compression):
+    name = "gzip"
+    file_extension = "tar.gz"
+
+    def __init__(self, command):
+        """
+
+        :param command: barman.fs.UnixLocalCommand
+        """
+        self.command = command
+
+    def uncompress(self, src, dst, exclude=None, include_args=None):
+        """
+
+        :param src: should specify if compression extension is included or not
+        :param dst: destination path
+        :param exclude: list of filepath in the archive to exclude from the extraction
+        :param include_args: list of filepath in the archive to extract.
+        :return:
+        """
+        if src is None or src == "":
+            raise ValueError("Source path should be a string")
+        if dst is None or dst == "":
+            raise ValueError("Destination path should be a string")
+        exclude = [] if exclude is None else exclude
+        exclude_args = []
+        for name in exclude:
+            exclude_args.append("--exclude")
+            exclude_args.append(name)
+        include_args = [] if include_args is None else include_args
+        args = ["xzf", src, "--directory", dst]
+        args.extend(exclude_args)
+        args.extend(include_args)
+        ret = self.command.cmd("tar", args=args)
+        out, err = self.command.get_last_output()
+        if ret != 0:
+            raise CommandFailedException(
+                "Error decompressing %s into %s: %s" % (src, dst, err)
+            )
+        else:
+            return self.command.get_last_output()
+
+    def get_file_content(self, filename, archive):
```

Comment:
Does this specific function get called anywhere? I'm having trouble finding a reference to it being called on `GZipCompression` (other than in the tests) and I'm wondering whether it is used at all, since by the end of the #555 PR we had switched to copying the config files from the recovery target directory rather than extracting them from the staged tarball.

Reply:
Thinking about this some more, perhaps we would be better off using this `get_file_content` method to read the backup_label, since we would then be relying only on external tools to handle compression-specific concerns rather than python libraries *and* external tools. Not something which needs action now but when we do add zstd and lz4 it might be a worthwhile change.